upload-notebooks/upload-notebook.py

```diff
@@ -0,0 +1,120 @@
+)
+parser.add_argument(
+    "notebook",
+    help="Notebook file to upload (extension .ipynb)",
+)
+parser.add_argument(
+    "--tiledb-cloud-token",
+    required=False,
+    default=os.environ.get("TILEDB_CLOUD_TOKEN"),
+    help="API token for TileDB Cloud account",
+)
+parser.add_argument(
+    "--tiledb-cloud-namespace",
+    required=False,
+    default=os.environ.get("TILEDB_CLOUD_NAMESPACE"),
+    help="Namespace for TileDB Cloud",
+)
+parser.add_argument(
+    "--tiledb-cloud-storage-path",
+    required=False,
+    default=os.environ.get("TILEDB_CLOUD_STORAGE_PATH"),
+    help="Storage path on TileDB Cloud",
+)
+parser.add_argument(
+    "--tiledb-cloud-storage-credential-name",
+    required=False,
+    default=os.environ.get("TILEDB_CLOUD_STORAGE_CREDENTIAL_NAME"),
+    help="Storage credentials to use on TileDB Cloud",
+)
+args = parser.parse_args()
+
+notebook = args.notebook
+token = args.tiledb_cloud_token
+namespace = args.tiledb_cloud_namespace
+storage_path = args.tiledb_cloud_storage_path
+storage_credential_name = args.tiledb_cloud_storage_credential_name
+
+# Verify input
+if not os.path.exists(notebook):
+    sys.stderr.write("Error: Notebook file does not exist: %s\n" % (notebook))
+    sys.exit(1)
+
+notebook_name, ext = os.path.splitext(notebook)
+notebook_name = os.path.basename(notebook_name)
+if ext != ".ipynb":
+    sys.stderr.write(
+        "Error: Notebook file extension must be '.ipynb', not '%s'\n" % (ext)
+    )
+    sys.exit(1)
+
+if token is None:
+    sys.stderr.write("Error: Missing API token\n")
+    sys.stderr.write(
+        "You must define the token as the env var TILEDB_CLOUD_TOKEN or pass it to --tiledb-cloud-token\n"
+    )
+    sys.exit(1)
+
+if namespace is None:
+    sys.stderr.write("Error: Missing namespace\n")
+    sys.stderr.write(
+        "You must define the namespace as the env var TILEDB_CLOUD_NAMESPACE or pass it to --tiledb-cloud-namespace\n"
+    )
+    sys.exit(1)
+
+# Login and upload
+tiledb.cloud.login(token=token)
+user = tiledb.cloud.user_profile()
+sys.stderr.write("Info: Logged into TileDB Cloud as %s\n" % (user.username))
+sys.stderr.write("Info: Notebook to upload is %s\n" % (notebook))
+sys.stderr.write("Info: Namespace is %s\n" % (namespace))
+if storage_path is not None:
+    sys.stderr.write("Info: Storage path is %s\n" % (storage_path))
+if storage_credential_name is not None:
+    sys.stderr.write(
+        "Info: Storage credential name is %s\n" % (storage_credential_name)
+    )
+
+try:
+    tiledb.cloud.upload_notebook_from_file(
+        ipynb_file_name=notebook,
+        namespace=namespace,
+        array_name=notebook_name,
+        storage_path=storage_path,
+        storage_credential_name=storage_credential_name,
+        on_exists=tiledb.cloud.notebook.OnExists.OVERWRITE,
+    )
+except tiledb.cc.TileDBError:
+    # If it's a new file, can't use OVERWRITE
```

Comment:
so eg
```
exists = tiledb.array_exists(storage_path)
...
   on_exists = OnExists.OVERWRITE if not exists else None
```

Reply:
Definitely. That's a much cleaner solution than relying on `try`/`except`

Reply:
Ok, two UX issues here - sorry. For the `array_exists` query, that's failing because the Cloud query requires a "context" for the cloud authentication. Try this:

```
# set context globally for the TileDB-Py session
tiledb.default_ctx(tiledb.cloud.Config())


# or set within this scope:

with tiledb.scope_ctx(tiledb.cloud.Ctx()):
    tiledb.array_exists("tiledb://jdblischak/cd88b458-007d-4974-bc57-40d28fbe688a")
```

Reply:
@ihnorton Thanks for the additional code. I can now detect the existing notebook locally using the TileDB URI:


```python
import os
import tiledb
import tiledb.cloud
token = os.environ.get("TILEDB_CLOUD_TOKEN")
tiledb.cloud.login(token=token)

with tiledb.scope_ctx(tiledb.cloud.Ctx()):
    exists = tiledb.array_exists("tiledb://jdblischak/cd88b458-007d-4974-bc57-40d28fbe688a")

exists
## True
```

Now that it's working locally, we need to decide how this will work in practice for users using this action in their GitHub Actions workflows. The action currently supports a TileDB token, not an AWS token. Is it possible for me to calculate the UUID used on TileDB Cloud from the name of the notebook file or its contents?

Reply:
> The action currently supports a TileDB token, not an AWS token.

We'll have to make the `tiledb.cloud.Ctx` usage conditional on whether the URI is `tiledb://`.

> Is it possible for me to calculate the UUID used on TileDB Cloud from the name of the notebook file or its contents?

No, no relation to the name or contents (it's generated based on a spec). I think we'll need to parametrize the target on a per-notebook basis.